### PR TITLE
[Wallet] Shielded transactions balance cache system.

### PR DESCRIPTION
**Pull request by furszy**
Up until now, one of the several capabilities that #1798 introduced was the wallet's shielded transactions balances calculation, a purely on-demand process (which is not so performant as txs balances are requested more than once during the entire software lifecycle). 
This PR is going further, introducing a new wallet balance cache system for shielded transactions. Covered by a good amount of units tests that can be run pre and post the new cache system.

This PR is obviously built on top of 1798. Part of the shielded GUI work that will be coming soon.

### CODE REVIEW

**Comment by furszy** [comment]
rebased on master now that #1798 got merged.

**Comment by furszy** [comment]
Pushed the fix for the lock held assertion issue on the new test cases. Travis happy now, ready for review.

**Review by random-zebra — APPROVED** [review]
utACK 9047f74e1cbb51783669e949e91087ae60f56327 (with few nits that can be ignored).

File: src/sapling/saplingscriptpubkeyman.cpp [hunk 1]
```diff
@@ -527,6 +527,34 @@ CAmount SaplingScriptPubKeyMan::GetCredit(const CWalletTx& tx, const isminefilte
     return nCredit;
 }
 
+CAmount SaplingScriptPubKeyMan::GetDebit(const CTransaction& tx, const isminefilter& filter)
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/sapling/saplingscriptpubkeyman.cpp:555]: yes.

**Review by Fuzzbawls — APPROVED** [review]
utACK 9047f74e1cbb51783669e949e91087ae60f56327

Also agree that the remaining nits can be done in their own PR

**Pull request merged**